Reactor Module 2025 Training material updates (#30071)
idaholab/moose: modules/doc/content/getting_started/examples_and_tutorials/tutorial04_meshing/step11_rgmb_het_hom.md

# Reactor Geometry Mesh Builder Example: Conversion of Heterogeneous to Homogeneous Sodium-Cooled Fast Reactor Core Mesh (ABTR)

> 

Reviewer: maybe add a note

this step is Griffin only, the regular MOOSE tutorial ends here

Reviewer 2: Most of this tutorial is MOOSE only (everything up to het mesh generation), only the last step (conversion from het to hom) is Griffin only

Reviewer: I agree but we should signal the end of the MOOSE-only content smh

Reviewer 2: I will add a comment about this

Reviewer 2: Added a comment to this file where the endpoint is

!listing reactor_examples/rgmb_abtr/rgmb_abtr_het_mesh.i
         id=tutorial04-rgmb_abtr-rmp
         caption=ABTR RGMB Reactor Mesh Parameters example.
         block=Mesh/rmp

## Pin structures using PinMeshGenerator

[PinMeshGenerator.md] defines the constituent pin structures used for stitching into assemblies. The pin pitch, number of azimuthal sectors, and geometry / region ID information about each ring, background, and duct region are specified here.

[PinMeshGenerator.md] is called multiple times to define the various pin structures (3 fuel pin types and 1 control pin type).

!listing reactor_examples/rgmb_abtr/rgmb_abtr_het_mesh.i
         id=tutorial04-rgmb_abtr-pin
         caption=ABTR RGMB pin example.
         block=Mesh/fuel_pin_1

Tips

> 

Reviewer 2: Added this to all Tips in this file

- Use a unique [PinMeshGenerator.md] block for each pin with a unique geometrical configuration and/or region ID composition
- region_ids is a 2-dimensional array containing region IDs (essentially materials). The first row of the array represents the 2D radial regions (from center of the pin to outermost region) for the bottom layer of the pin. Each subsequent row assigns IDs on another axial level, from bottom to top. In this case, each pin has 3 radial regions (fuel, clad, background), this array is a column pertaining to each axial level of the pin assembly.
- While the mesh is still 2D during this step, the axially dependent region IDs are stored for later use during the extrusion step.

> 

Reviewer: ```suggestion
!alert-end!

```

## Assembly structures using AssemblyMeshGenerator

[AssemblyMeshGenerator.md] takes the pin types previously defined and places them into a regular hexagonal grid. Additionally, coolant and duct regions need to be added around the pins in order to create the assembly geometry.

[AssemblyMeshGenerator.md] is called multiple times to define the various heterogeneous assemblies (3 fuel assemblies and 1 control assembly).

!listing reactor_examples/rgmb_abtr/rgmb_abtr_het_mesh.i
         id=tutorial04-rgmb_abtr-het-assembly
         caption=ABTR RGMB heterogeneous assembly example.
         block=Mesh/fuel_assembly_1

> Tips

Reviewer: same format could look beter

- Use a unique [AssemblyMeshGenerator.md] block for each assembly with a unique geometrical configuration, region ID composition, and/or inventory of pin structures

> - background_region_ids is a 1-dimensional array containing region IDs for each axial layer of the background region. duct_region_ids is a 2-dimensional array containing regions IDs for the duct region. The first row of the array represents the 2D duct regions (from innermost to outermost duct region) for the bottom layer of the assembly. Each subsequent row assigns IDs on another axial level, from bottom to top.

Reviewer: ```suggestion
- [!param](/Mesh/AssemblyMeshGenerator/background_region_ids) is a 1-dimensional array containing region IDs for each axial layer of the background region. duct_region_ids is a 2-dimensional array containing regions IDs for the duct region. The first row of the array represents the 2D duct regions (from innermost to outermost duct region) for the bottom layer of the assembly. Each subsequent row assigns IDs on another axial level, from bottom to top.
```
same for any other param please. it prevents deprecation of the training content

Reviewer 2: Updated, fixed other instances too. Let me know if I missed anything

- While the mesh is still 2D during this step, the axially dependent region IDs are stored for later use during the extrusion step.

## Homogeneous assembly structures using PinMeshGenerator

In order to define homogeneous assembly structures to stitch into the core, we use [PinMeshGenerator.md] once again ([AssemblyMeshGenerator.md] is only used for structures that consist of lattices of pins).

- +To define single assemblies directly with PinMeshGenerators for stitching with [CoreMeshGenerator.md]+, [PinMeshGenerator.md] is used with [!param](/Mesh/PinMeshGenerator/use_as_assembly) set to `true`.

> - In addition, [!param](/Mesh/PinMeshGenerator/homogenized) = `true` is used to indicate that this region is homogenized and [SimpleHexagonGenerator.md] should be called to discretize the assembly instead of [PolygonConcentricCircleMeshGenerator.md].

Reviewer: the SimpleHexagonGenerator does not actually show up in the input ?

this could confuse users. I would not mention it

Reviewer 2: Done

!listing reactor_examples/rgmb_abtr/rgmb_abtr_het_mesh.i
         id=tutorial04-rgmb_abtr-hom-assembly
         caption=ABTR RGMB homogeneous assembly example.
         block=Mesh/reflector_assembly

> Tips

Reviewer: same comments

- For homogenized assemblies, each assembly has only 1 radial region, so each column in `region_ids` pertains to the region ID of each axial level of the homogenized assembly.
- While the mesh is still 2D during this step, the axially dependent region IDs are stored for later use during the extrusion step.

## Heterogeneous core using CoreMeshGenerator

Now that all heterogeneous and homogeneous assemblies have been defined, they are placed into a lattice using [CoreMeshGenerator.md]. While [CoreMeshGenerator.md] still requires a perfect hexagonal pattern like [PatternedHexMeshGenerator.md], it automatically handles dummy assembly creation and deletion. The user need only provide a fake mesh input reference `dummy` (this object has not been actually created) and tell [CoreMeshGenerator.md] through the [!param](/Mesh/CoreMeshGenerator/dummy_assembly_name) parameter that the mesh input called `dummy` is a dummy assembly (empty space). The dummy assemblies will be created and deleted behind the scenes with no effort from the user.

> - Since we want to extrude the 2D core, we use the [!param](/Mesh/CoreMeshGenerator/extrude)=`true` parameter within [CoreMeshGenerator.md]. The step simultaneously extrudes the geometry and applies the regions IDs to all axial layers as defined in the [PinMeshGenerator.md] objects.

Reviewer: it also processes the assembly MG ducts and background axial parameters too?

Reviewer 2: Updated to include references to both PMG and AMG objects

> - Behind the scenes, extra element IDs `assembly_id` and `plane_id` are automatically applied.

Reviewer: ```suggestion
- Behind the scenes, extra element IDs `assembly_id` and `plane_id` are automatically added to the relevant elements
```

- Since [!param](/Mesh/ReactorMeshParams/flexible_assembly_stitching) = `true` in [ReactorMeshParams.md], the different assembly types are stitched together without hanging nodes.

!listing reactor_examples/rgmb_abtr/rgmb_abtr_het_mesh.i
         id=tutorial04-rgmb_abtr-het_core
         caption=ABTR RGMB core example.
         block=Mesh/het_core

!alert tip
[!param](/Mesh/CoreMeshGenerator/extrude) = `true` indicates extrusion should happen at this step (this should only be set to true once in the entire input file)

!media tutorial04_meshing/rgmb_abtr_het_core_eeid.png
       id=tutorial04-rgmb_abtr_het_core
       caption=RGMB-generated 3D heterogeneous core mesh, colored by various IDs defined on the mesh.
       style=width:75%;display:block;margin-left:auto;margin-right:auto;

> ## Equivalent homogeneous core using Griffin's EquivalentCoreMeshGenerator

Reviewer: or should the tutorial end right here for MOOSE-only?

Reviewer: (the griffin content can stay for convenience)

Reviewer 2: Added a description here to indicate end of MOOSE content and beginning of Griffin content

## Homogeneous assembly structures using PinMeshGenerator

In order to define homogeneous assembly structures to stitch into the core, we use [PinMeshGenerator.md] once again ([AssemblyMeshGenerator.md] is only used for structures that consist of lattices of pins).

- +To define single assemblies directly with PinMeshGenerators for stitching with [CoreMeshGenerator.md]+, [PinMeshGenerator.md] is used with [!param](/Mesh/PinMeshGenerator/use_as_assembly) set to `true`.
- In addition, [!param](/Mesh/PinMeshGenerator/homogenized) = `true` is used to indicate that this region is homogenized.

!listing reactor_examples/rgmb_abtr/rgmb_abtr_het_mesh.i
         id=tutorial04-rgmb_abtr-hom-assembly
         caption=ABTR RGMB homogeneous assembly example.
         block=Mesh/reflector_assembly

!alert! note title=Tips

- For homogenized assemblies, each assembly has only 1 radial region, so each column in `region_ids` pertains to the region ID of each axial level of the homogenized assembly.
- While the mesh is still 2D during this step, the axially dependent region IDs are stored for later use during the extrusion step.

!alert-end!

## Heterogeneous core using CoreMeshGenerator

> At this point, we have defined a 3D heterogeneous ABTR Now that all heterogeneous and homogeneous assemblies have been defined, they are placed into a lattice using [CoreMeshGenerator.md]. While [CoreMeshGenerator.md] still requires a perfect hexagonal pattern like [PatternedHexMeshGenerator.md], it automatically handles dummy assembly creation and deletion. The user need only provide a fake mesh input reference `dummy` (this object has not been actually created) and tell [CoreMeshGenerator.md] through the [!param](/Mesh/CoreMeshGenerator/dummy_assembly_name) parameter that the mesh input called `dummy` is a dummy assembly (empty space). The dummy assemblies will be created and deleted behind the scenes with no effort from the user.

Reviewer 2: Actually removing this line because a full 3-D mesh hasn't been generated by this point yet
